src/track/beatutils.cpp

```diff
@@ -12,37 +12,314 @@
 #include <QMap>
 
 #include "track/beatutils.h"
+#include "track/beatstats.h"
 #include "util/math.h"
 
+namespace {
+
 // we are generous and assume the global_BPM to be at most 0.05 BPM far away
 // from the correct one
-#define BPM_ERROR 0.05
+constexpr double kMaxBpmError = 0.05;
+constexpr double kMaxSecsPhaseError = 0.025; //25ms
 
 // the raw beatgrid is divided into blocks of size N from which the local bpm is
 // computed. Tweaked from 8 to 12 which improves the BPM accuracy for 'problem songs'.
-#define N 12
+constexpr int kBeatsToCountTempo = 12;
 
 static bool sDebug = false;
 
 const double kCorrectBeatLocalBpmEpsilon = 0.05; //0.2;
 const int kHistogramDecimalPlaces = 2;
 const double kHistogramDecimalScale = pow(10.0, kHistogramDecimalPlaces);
 const double kBpmFilterTolerance = 1.0;
+constexpr double kMaxDiffSameBpm = 0.25;
+
+bool checkIfErrorGrowing(double previous, double current) {
+
+    if (current < 0 && previous > 0) {
+        return false;
+    } else if (current > 0 && previous < 0) {
+        return true;
+    } else {
+        return fabs(current) > fabs(previous)
+                ? true : false;
+    }
+}
+
+} //namespece
+
+QMap<int, double> BeatUtils::findTempoChanges(
+        QMap<double, int> tempoFrequency, QList<double> tempoList) {
+
+    auto sortedTempoList = tempoList;
+    std::sort(sortedTempoList.begin(), sortedTempoList.end());
+    // We have to make sure we have odd numbers
```

Comment:
it's 2 lines below. 
Should just make all the comment together, right?

Reply:
Not really. 
"A median filter need an odd number to be able to return a single center value."

Thinking about this, I think it is wrong to randomly drop a value. What if you accidentally drop the median value? 
I think it is better to not drop a value, but return one of the candidates or the average values of both.

Reply:
The whole point is that we can not use the average of the values because it might be an outlier from our frequencyOfTempos table. 
Returning both candidates will work though. But honestly, I don`t think we are likely to remove the median. The median is very likely to be a repeated value in the list. Also we just use it as a rough guess for the tempo to have something to start working with, if we actually remove the true median, the "next median" should be an equally good rough guess

Reply:
Ok, drop a note of that into the source.